modules/tfm/tfm/boards/common/nrf_provisioning.c

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include "tfm_plat_provisioning.h"
+#include "tfm_plat_otp.h"
+#include "tfm_hal_platform.h"
+#include "tfm_attest_hal.h"
+#include "hw_unique_key.h"
+#include "nrfx_nvmc.h"
+#include <nrf_cc3xx_platform.h>
+#include <nrf_cc3xx_platform_identity_key.h>
+
+int tfm_plat_provisioning_is_required(void)
+{
+    enum tfm_security_lifecycle_t lcs = tfm_attest_hal_get_security_lifecycle();
+
+    return lcs != TFM_SLC_SECURED;
+}
+
+enum tfm_plat_err_t tfm_plat_provisioning_perform(void)
+{
+    enum tfm_security_lifecycle_t lcs = tfm_attest_hal_get_security_lifecycle();
+
+    /*
+     * Provisioning in NRF defines has two steps, the first step is to execute
+     * the identity_key_generation sample. When this sample is executed it will
+     * always set the lifecycle state to PROVISIONING. This is a requirement for
+     * the TF-M provisioning to be completed so we don't accept any other
+     * lifecycle state here.
+     */
+    if(lcs != TFM_SLC_PSA_ROT_PROVISIONING) {
+        return TFM_PLAT_ERR_SYSTEM_ERR;
+    }
+
+    /* The Hardware Unique Keys should be already written */
+    if(!hw_unique_key_are_any_written()) {
+        return TFM_PLAT_ERR_SYSTEM_ERR;
+    }
+
+    /* The Initial Attestation key should be already written */
+    if(!nrf_cc3xx_platform_identity_key_is_stored(NRF_KMU_SLOT_KIDENT)) {
+        return TFM_PLAT_ERR_SYSTEM_ERR;
+    }
+
+    /*
+     * We don't need to make sure that the validation key is written here since we expect
+     * that secure boot is already enabled at this stage
+     */
+
+    /* Disable debugging in the UICR register */
+    bool approt_writable = nrfx_nvmc_word_writable_check((uint32_t)&NRF_UICR->APPROTECT, UICR_APPROTECT_PALL_Protected);
+    approt_writable &= nrfx_nvmc_word_writable_check((uint32_t)&NRF_UICR->SECUREAPPROTECT, UICR_SECUREAPPROTECT_PALL_Protected);
+
+    if(approt_writable){
+        nrfx_nvmc_word_write((uint32_t)&NRF_UICR->APPROTECT, UICR_APPROTECT_PALL_Protected);
+        nrfx_nvmc_word_write((uint32_t)&NRF_UICR->SECUREAPPROTECT, UICR_SECUREAPPROTECT_PALL_Protected);
+    } else {
+        return TFM_PLAT_ERR_SYSTEM_ERR;
+    }
+
+    /* Transition to the SECURED lifecycle state */
+    if (tfm_attest_update_security_lifecycle_otp(TFM_SLC_SECURED) != 0){
+        return TFM_PLAT_ERR_SYSTEM_ERR;
+    }
+
+    lcs = tfm_attest_hal_get_security_lifecycle();
+    if (lcs != TFM_SLC_SECURED){
+        return TFM_PLAT_ERR_SYSTEM_ERR;
+    }
+
+    /* Perform a mandatory reset since we switch to an attestable LCS state */
+    NVIC_SystemReset();
+
+    /*
+     * We should never return from this function, a reset should be triggered
+     * before we reach this point. Returning an error to signal that something
+     * is wrong if we reached here.
+     */
+    return TFM_PLAT_ERR_SYSTEM_ERR;
+}
+
+void tfm_plat_provisioning_check_for_dummy_keys(void)
+{
```

Comment:
Should we still check for the provisioned dummy data?
Maybe someone did DFU from debug-TF-M with dummy keys to production build?

Reply:
Doesn't sounds like a stupid idea, to still check for dummy keys

Reply:
TF-M stores the provisioned dummy key in flash I believe, and can easily read it out.

But we store it in encrypted form in the KMU.

The code size / runtime overhead of invoking identity_key_read to decrypt and check the dummy key is not worth the security it brings IMHO.

@joerchan : Okay to resolve this as won't fix?

Reply:
Did a prototype.

It's surprisingly expensive, even though the API we call is already present.

180 bytes of RAM

264 bytes of flash

Reply:
TF-M has no infrastructure to add custom defines to easily make some code conditional, so I suggest not supporting this.

Reply:
Pushed a commit for checkig dummy keys, 190 bytes of flash.

I suppose configuring SPMLOG_ERR to be disabled will cut away this resource usage  so it's okay to spend some bytes here.